### PR TITLE
Update etcd, improve nixos module, fix nixos tests

### PR DESCRIPTION
**Pull request by offlinehacker**
###### Motivation for this change

For kubernetes update i'm working on
###### Things done
- [x] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/tests/etcd-cluster.nix [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    [ v3_req ]
+    basicConstraints = CA:FALSE
+    keyUsage = digitalSignature, keyEncipherment
+    extendedKeyUsage = clientAuth
+    EOF
+
+    openssl genrsa -out $out/etcd-client-key.pem 2048
+    openssl req -new -key $out/etcd-client-key.pem -out etcd-client.csr -subj "/CN=etcd-client" -config client-openssl.cnf
+    openssl x509 -req -in etcd-client.csr -CA $out/ca.pem -CAkey $out/ca-key.pem -CAcreateserial -out $out/etcd-client.pem -days 365 -extensions v3_req -extfile client-openssl.cnf
+  '';
```
Comment (grahamc) [inline nixos/tests/etcd-cluster.nix]: I've found it helpful to break these out in to separate buildCommand calls so it is clearer what depends on what and how it fits together. Would you be interested in a patch to your PR which breaks each of these in to its own derivation? Does that seem like overkill?

Comment (grahamc) [inline nixos/tests/etcd-cluster.nix]: By no means do you have to use this, but I quite like the pattern in the linked commit. I've found people are less scared of it when each command is split up with these very well defined inputs and outputs. That said, a test inside nixos might not be the right place for me to grind that axe :)

https://github.com/NixOS/nixpkgs/commit/db6bb8de5346bccd5307cc6d3598b76d6278bf63

File: nixos/modules/services/misc/etcd.nix [hunk 2]
```diff
@@ -99,15 +147,22 @@ in {
...
+      environment = (filterAttrs (n: v: v != null) {
         ETCD_NAME = cfg.name;
         ETCD_DISCOVERY = cfg.discovery;
         ETCD_DATA_DIR = cfg.dataDir;
         ETCD_ADVERTISE_CLIENT_URLS = concatStringsSep "," cfg.advertiseClientUrls;
         ETCD_LISTEN_CLIENT_URLS = concatStringsSep "," cfg.listenClientUrls;
         ETCD_LISTEN_PEER_URLS = concatStringsSep "," cfg.listenPeerUrls;
         ETCD_INITIAL_ADVERTISE_PEER_URLS = concatStringsSep "," cfg.initialAdvertisePeerUrls;
-      } // (optionalAttrs (cfg.discovery == ""){
+        ETCD_PEER_TRUSTED_CA_FILE = cfg.peerTrustedCaFile;
```
Comment (teh) [inline nixos/modules/services/misc/etcd.nix]: Are certs now required? What happens if I leave the default `null` in? Will the env var be unset?

Comment (offlinehacker) [inline nixos/modules/services/misc/etcd.nix]: It filters environment variables which have value of null, so all good

File: nixos/tests/etcd-cluster.nix (hunk 1 above)
Comment (offlinehacker) [inline nixos/tests/etcd-cluster.nix]: Ok, let's do this, i have no preference, so i'm fine with suggestion.

**Comment by fpletz** [comment]
Is this ready to merge?

**Comment by offlinehacker** [comment]
Yes, i haven't tested iwhat happens if updating nixos from older etcd version, but it should work out of the box, but for etcd, you should have backup in any case.

**Pull request merged**